src/components/shop-app.js

```diff
@@ -0,0 +1,507 @@
+    page: String,
+
+    offline: Boolean,
+
+    meta: Object,
+    
+    modalOpened: Object,
+
+    categories: Object,
+
+    categoryName: String,
+
+    drawerOpened: Boolean,
+
+    _a11yLabel: String,
+
+    _smallScreen: Boolean,
+
+    _snackbarOpened: Boolean,
+
+    loadComplete: Boolean
+  }}
+
+  _propertiesChanged(props, changed, oldProps) {
+    if (changed) {
+      if ('categoryName' in changed) {
+        this._categoryNameChanged(props.categoryName, oldProps.categoryName);
+      }
+      if ('page' in changed) {
+        this._pageChanged(props.page, oldProps.page);
+      }
+      if ('offline' in changed) {
+        this._offlineChanged(props.offline, oldProps.offline);
+      }
+      if ('meta' in changed) {
+        this._metaChanged(props.meta, oldProps.meta);
+      }
+    }
+    super._propertiesChanged(props, changed, oldProps);
+  }
+
+  constructor() {
+    super();
+
+    installRouter(() => this._updateLocation());
+    installNetwork(store);
+    store.dispatch(fetchCategories());
+  }
+
+  stateChanged() {
+    const state = store.getState();
+
+    this._category = currentCategorySelector(state);
+    this._item = currentItemSelector(state);
+    let page = pageSelector(state);
+    switch (page) {
+      case 'list':
+        if (!this._category) page = '404';
+        break;
+      case 'detail':
+        if (!this._category || !this._item) page = '404';
+        break;
+      case 'home':
+      case 'cart':
+      case 'checkout':
+        break;
+      default:
+        page = '404';
+    }
+
+    this.categories = Object.values(state.categories);
+    this.categoryName = this._category ? this._category.name : null;
+    this.meta = state.meta;
+    this.modalOpened = state.modal;
+    this.offline = !state.network.online;
+    this.page = page;
+    this._a11yLabel = state.announcer.label;
+  }
+
+  ready() {
+    super.ready();
+    // Custom elements polyfill safe way to indicate an element has been upgraded.
+    this.removeAttribute('unresolved');
+    // listen for custom events
+    // this.addEventListener('add-cart-item', (e)=>this._onAddCartItem(e));
+    // this.addEventListener('set-cart-item', (e)=>this._onSetCartItem(e));
+    // this.addEventListener('clear-cart', (e)=>this._onClearCart(e));
+    // this.addEventListener('change-section', (e)=>this._onChangeSection(e));
+    // this.addEventListener('announce', (e)=>this._onAnnounce(e));
+    // this.addEventListener('dom-change', (e)=>this._domChange(e));
+    // this.addEventListener('show-invalid-url-warning', (e)=>this._onFallbackSelectionTriggered(e));
+
+    const mq = window.matchMedia('(max-width: 767px)');
+    mq.addListener(_ => this._smallScreen = mq.matches);
+    this._smallScreen = mq.matches;
+  }
+
+  _categoryNameChanged(categoryName, oldCategoryName) {
+    store.dispatch(fetchCategoryItems(this._category));
+
+    // Reset the list view scrollTop if the category changed.
+    this._listScrollTop = 0;
```

Comment:
The scrolling is fairly complicated and could be part of the app state. It's a rendering concept so we should try to do this work in render although that may be a bit tricky.